src/clj/rems/api/services/dependencies.clj

```diff
@@ -0,0 +1,131 @@
+(ns rems.api.services.dependencies
+  "Tracking dependencies between catalogue items, resources, forms, workflows and licenses."
+  (:require [medley.core :refer [map-vals]]
+            [rems.db.catalogue :as catalogue]
+            [rems.db.form :as form]
+            [rems.db.licenses :as licenses]
+            [rems.db.resource :as resource]
+            [rems.db.workflow :as workflow]))
+
+(defn enrich-dependency [dep]
+  (cond
+    (:license/id dep) (licenses/get-license (:license/id dep))
+    (:resource/id dep) (resource/get-resource (:resource/id dep))
+    (:workflow/id dep) (workflow/get-workflow (:workflow/id dep))
+    (:catalogue-item/id dep) (catalogue/get-localized-catalogue-item (:catalogue-item/id dep))
+    (:form/id dep) (form/get-form-template (:form/id dep))
+    true (assert false dep)))
+
+(defn- list-dependencies []
+  (concat
+
+   (for [res (resource/get-resources {})
+         license (:licenses res)]
+     {:from {:resource/id (:id res)} :to {:license/id (:id license)}})
+
+   (for [cat (catalogue/get-localized-catalogue-items {:archived true})
+         dep [{:form/id (:formid cat)} {:resource/id (:resource-id cat)} {:workflow/id (:wfid cat)}]]
+     {:from {:catalogue-item/id (:id cat)} :to dep})
+
+   (for [workflow (workflow/get-workflows {})
+         dep (concat
+              (mapv (fn [lic] {:license/id (:id lic)}) (:licenses workflow))
+              (:forms (:workflow workflow)))]
+     {:from {:workflow/id (:id workflow)} :to dep})))
+
+(defn- add-status-bits [dep]
+  (merge dep
+         (select-keys (enrich-dependency dep) [:archived :enabled])))
+
+(defn- only-id [item]
+  (select-keys item [:resource/id :license/id :catalogue-item/id :form/id :workflow/id]))
+
+(defn- list-to-maps [lst]
+  {:dependencies
+   (map-vals (comp set (partial map :to)) (group-by :from lst))
+   :reverse-dependencies
```

Comment:
Could better names for these be `depends-on` and `used-by` or something along those lines?

Reply:
I considered that, but dependencies/reverse dependencies are standard terms and also match the name of the ns nicely.

Reply:
One thing to consider would be also to avoid using the map directly so you can arrange it differently after the fact. Of course if all the uses are in this ns and not public it'll be easy anyway.

Reply:
I was mainly thinking about how to read the uses of the map.